lamindb/curators/__init__.py

```diff
@@ -703,22 +730,186 @@ def save_artifact(
+        if not data_is_spatialdata(self._dataset):
+            raise InvalidArgument("dataset must be SpatialData-like.")
+        if schema.otype != "SpatialData":
+            raise InvalidArgument("Schema otype must be 'SpatialData'.")
+
+        # in form of {table_key: var_field}
+        self._var_fields: dict[str, FieldAttr] = {}
+        # in form of {table_key: categoricals}
+        self._categoricals: dict[str, dict[str, FieldAttr]] = {}
+        for slot, slot_schema in schema.slots.items():
+            # Assign to _slots
+            if ":" in slot:
+                table_key, table_slot = slot.split(":")
+                schema_dataset = self._dataset.tables.__getitem__(table_key)
+            # sample metadata (does not have a `:` separator)
+            else:
+                table_key = None
+                table_slot = slot
+                schema_dataset = self._dataset.get_attrs(
+                    key=sample_metadata_key, return_as="df", flatten=True
+                )
+
+            self._slots[slot] = DataFrameCurator(
+                (
+                    schema_dataset.__getattribute__(table_slot).T
+                    if table_slot == "var"
+                    else schema_dataset.__getattribute__(table_slot)
+                    if table_slot != sample_metadata_key
+                    else schema_dataset  # just take the schema_dataset if it's the sample metadata key
+                ),
+                slot_schema,
+            )
+
+            _assign_var_fields_categoricals_multimodal(
+                table_key=table_key,
+                table_slot=table_slot,
+                slot=slot,
+                slot_schema=slot_schema,
+                var_fields=self._var_fields,
+                categoricals=self._categoricals,
+                slots=self._slots,
+            )
+
+        # for consistency with BaseCatManager
+        self._columns_field = self._var_fields
+
+    @property
+    @doc_args(SLOTS_DOCSTRING)
+    def slots(self) -> dict[str, DataFrameCurator]:
+        """{}"""  # noqa: D415
+        return self._slots
+
+    @doc_args(VALIDATE_DOCSTRING)
+    def validate(self) -> None:
+        """{}"""  # noqa: D415
+        for _, curator in self._slots.items():
+            curator.validate()
+
+    @doc_args(SAVE_ARTIFACT_DOCSTRING)
+    def save_artifact(
+        self,
+        *,
+        key: str | None = None,
+        description: str | None = None,
+        revises: Artifact | None = None,
+        run: Run | None = None,
+    ):
+        """{}"""  # noqa: D415
+        if not self._is_validated:
+            self.validate()
+
+        return save_artifact(  # type: ignore
+            self._dataset,
+            key=key,
+            description=description,
+            fields=self._categoricals,
+            index_field=self._var_fields,
+            artifact=self._artifact,
+            revises=revises,
+            run=run,
+            schema=self._schema,
+        )
```

Comment:
This and also the `MudataCurator.save_artifact` can both go into `SlotsCurator` once you move `self._var_fields` and `self._categoricals` initialization as well. This whole thing can be much more compact.

Reply:
Moved now and noted that this is MDC and SDC specific.